### PR TITLE
feat: generic consent management

### PR DESCRIPTION
**Pull request by saikumarrs**
# Description

Add support for generic consent management.

## Linear Ticket

https://linear.app/rudderstack/issue/SDK-663/add-the-new-consent-managment-module-in-the-rudder-server

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Enhanced consent management logic with new data structures for improved handling of user consents.
  - Introduced new methods for retrieving consent data from various consent management providers.

- **Improvements**
  - Updated existing consent filtering logic to support both legacy and new consent management systems.
  - Refined test functions to align with the updated consent management logic and added new test cases for comprehensive coverage.

- **Refactor**
  - Renamed key functions to better reflect their purpose in the context of consent management.
  - Restructured test functions for improved clarity and maintainability.

- **Documentation**
  - Corrected a typographical error in the code comments for better code documentation.

- **Bug Fixes**
  - Adjusted function signatures and parameters to ensure consistency and correct handling of consent information.

### CODE REVIEW

**Review by cisse21 — APPROVED** [review]

**Review by cisse21 — CHANGES_REQUESTED** [review]
Requested few changes with regards to implementation. Also we can get all the consent related functions into a separate file which would make it easier to test

File: processor/processor.go [hunk 1]
```diff
@@ -2918,17 +3005,37 @@ func (proc *Handle) isDestinationAvailable(event types.SingularEventT, sourceId
...
+		return consentStr, consentStr != ""
+	})
+}
+
+func getConsentManagementInfo(se types.SingularEventT) ConsentManagementInfo {
+	sanitizedAllowedConsentIds := make([]string, 0)
+	sanitizedDeniedConsentIds := make([]string, 0)
+	provider := ""
+	resolutionStrategy := ""
+	if consentManagement, ok := misc.MapLookup(se, "context", "consentManagement").(map[string]interface{}); ok {
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply processor/processor.go]: Addressed.

File: processor/processor.go [hunk 2]
```diff
@@ -738,18 +751,62 @@ func getConsentCategories(dest *backendconfig.DestinationT) []string {
...
+
+	consentManagementConfig := dest.Config["consentManagement"].([]map[string]interface{})
+	if len(consentManagementConfig) == 0 {
+		return consentManagementData
+	}
+
+	for _, providerConfig := range consentManagementConfig {
+		provider, _ := providerConfig["provider"].(string)
+		resolutionStrategy, _ := providerConfig["resolutionStrategy"].(string)
+		consentsConfig, _ := providerConfig["consents"].([]interface{})
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply processor/processor.go]: Addressed.

File: processor/consentmanagementfilter/consentmanagementfilter.go
Comment (cisse21) [inline processor/consentmanagementfilter/consentmanagementfilter.go:1]: We would need some unit tests for this file too

File: processor/consentmanagementfilter/consentmanagementfilter.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	return genericConsentManagementData
+}
+
+func getValidConsents(consents []string) []string {
+	return lo.FilterMap(consents, func(consent string, _ int) (string, bool) {
+		return consent, consent != ""
+	})
+}
```
Comment (cisse21) [inline processor/consentmanagementfilter/consentmanagementfilter.go:105]:
```suggestion
```
Don't think we would need a function for this and can be done inline

File: processor/consentmanagementfilter/consentmanagementfilter.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		}
+
+		err := jsonfast.Unmarshal(consentManagementStr, &consentManagementInfo)
+		if err != nil {
+			return consentManagementInfo
+		}
+	}
+
+	consentManagementInfo.AllowedConsentIds = getValidConsents(consentManagementInfo.AllowedConsentIds)
+	consentManagementInfo.DeniedConsentIds = getValidConsents(consentManagementInfo.DeniedConsentIds)
```
Comment (cisse21) [inline processor/consentmanagementfilter/consentmanagementfilter.go:122]:
```suggestion
	consentManagementInfo.AllowedConsentIds = lo.FilterMap(consentManagementInfo.AllowedConsentIds, func(consent string, _ int) (string, bool) {
		return consent, consent != ""
	})
	consentManagementInfo.DeniedConsentIds = lo.FilterMap(consentManagementInfo.DeniedConsentIds, func(consent string, _ int) (string, bool) {
		return consent, consent != ""
	})
```

File: processor/processor.go [hunk 5]
```diff
@@ -2875,13 +2876,33 @@ func (proc *Handle) filterDestinations(
...
 	}
+
 	return lo.Filter(dests, func(dest backendconfig.DestinationT, _ int) bool {
-		if consentCategories := proc.getConsentCategories(dest.ID); len(consentCategories) > 0 {
-			return len(lo.Intersect(consentCategories, deniedCategories)) == 0
+		// This field differentiates legacy and generic consent management
+		if consentManagementInfo.Provider != "" {
+			if cmpData := proc.getGCMData(dest.ID, consentManagementInfo.Provider); len(cmpData.Consents) > 0 {
+
+				var finalResolutionStrategy string = consentManagementInfo.ResolutionStrategy
```
Comment (cisse21) [inline processor/processor.go:2889]:
```suggestion
				var finalResolutionStrategy = consentManagementInfo.ResolutionStrategy
```

**Review by cisse21 — CHANGES_REQUESTED** [review]

**Review by cisse21 — APPROVED** [review]
LGTM

**Review by cisse21 — APPROVED** [review]

File: processor/processor.go [hunk 6]
```diff
@@ -817,8 +819,14 @@ func (proc *Handle) backendConfigSubscriber(ctx context.Context) {
 					for j := range source.Destinations {
 						destination := &source.Destinations[j]
 						destinationIDtoTypeMap[destination.ID] = destination.DestinationDefinition.Name
-						oneTrustConsentCategoriesMap[destination.ID] = oneTrustConsentCategories(destination)
-						ketchConsentCategoriesMap[destination.ID] = ketchConsentCategories(destination)
+						oneTrustConsentCategoriesMap[destination.ID] = GetOneTrustConsentCategories(destination)
+						ketchConsentCategoriesMap[destination.ID] = GetKetchConsentCategories(destination)
+
+						err := error(nil)
```
Comment (Sidddddarth) [inline processor/processor.go:825]: this is causing a linter issue
we can declare it up there with other `var`s

Comment (saikumarrs) [reply processor/processor.go]: Thank you. I've addressed it.

File: processor/consent.go [hunk 7]
```diff
@@ -1,59 +1,131 @@
...
 	"github.com/rudderlabs/rudder-server/utils/types"
 )
 
-// filterDestinations filters destinations based on consent categories, supports oneTrustCookieCategories and ketchConsentPurposes
-func (proc *Handle) filterDestinations(event types.SingularEventT, destinations []backendconfig.DestinationT) []backendconfig.DestinationT {
-	// If there are no deniedConsentIds, then return all destinations
-	deniedCategories := deniedConsentCategories(event)
-	if len(deniedCategories) == 0 {
+type ConsentManagementInfo struct {
+	DeniedConsentIds   []string `json:"deniedConsentIds"`
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply processor/consent.go]: Addressed.

File: processor/consent.go [hunk 8]
```diff
@@ -1,59 +1,131 @@
...
+	AllowedConsentIds  []string `json:"allowedConsentIds"`
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply processor/consent.go]: Addressed.

File: processor/consent.go [hunk 9]
```diff
@@ -1,59 +1,131 @@
...
+Filters and returns destinations based on the consents configured for the destination and the user consents present in the event.
+
+Supports legacy and generic consent management.
+*/
+func (proc *Handle) getConsentFilteredDestinations(event types.SingularEventT, destinations []backendconfig.DestinationT) []backendconfig.DestinationT {
+	// If the event does not have denied consent IDs, do not filter any destinations
+	consentManagementInfo, err := getConsentManagementInfo(event)
+	if err != nil {
+		// Log the error for debugging purposes
+		proc.logger.Error(err)
```
Comment (achettyiitr) [inline processor/consent.go:45]: We can use structured logging here. Same for the other logging as well.
```suggestion
		proc.logger.Errorw("failed to get consent management info", "error", err.Error())
```

**Comment by achettyiitr** [comment]
Approving with minor comments left.

**Pull request merged**